### PR TITLE
Add a TimeZone interface to moment core

### PR DESCRIPTION
**Pull request by timrwood**
[Rendered](https://github.com/timrwood/moment-rfcs/blob/time-zone-interface/text/0000-time-zone-interface.md)

### CODE REVIEW

File: text/0000-time-zone-interface.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+4. We hack around the moment constructor in moment-timezone in a way that makes it impossible to parse time only input correctly. https://github.com/moment/moment-timezone/issues/297
+
+5. We have forked logic for local and utc mode getters and setters. `Date#getMonth` vs `Date#getUTCMonth`.
+
+# Detailed design
+
+To solve this, we could introduce an interface for a `TimeZone` in moment. Moment would provide 2 classes that implement the `TimeZone` interface, `LocalTimeZone` and `FixedOffsetTimeZone`. Moment timezone would provide a `TZDBTimeZone` class that implements the `TimeZone` interface.
+
+Each moment that was created would have a time zone. Calls to `moment()` would have a `LocalTimeZone` instance, calls to `moment.utc()` would have a `FixedOffsetTimeZone` instance, and
```
Comment (ichernev) [inline text/0000-time-zone-interface.md]: Shouldn't this be present tense?

```
Each moment that is created ...
```

File: text/0000-time-zone-interface.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+Sample implementations:
+
+```
+FixedOffsetTimeZone.prototype.parse = function (parts, config) {
+	return this.presetOffset;
+};
+
+LocalTimeZone.prototype.parse = function (parts, config) {
+	var parsed = new Date(parts[0], parts[1], parts[2], parts[3], parts[4];
```
Comment (ichernev) [inline text/0000-time-zone-interface.md]: missing closing `)`.

**Comment by ichernev** [comment]
I think after we introduce `TimeZone` we can safely deprecate whatever doesn't make too much sense.

Having `moment.defaultZone` also makes sense.

I think the current proposal is missing interface for parsing an offset. Or the current fixed offset will be special-case handled by moment, because only fixed-offset needs parsing?

As for the api itself -- if you manage to implement it, its good enough. Otherwise the spec can change to accommodate for the code.

**Comment by joetidee** [comment]
Is the timezone interface going to be released any time soon?

**Comment by timrwood** [comment]
@joetidee, I have started implementing at https://github.com/moment/moment/pull/3134, you can follow along there.

**Comment by timrwood** [comment]
I updated this RFC with findings from the initial implementation spike.

**Pull request closed without merging**